### PR TITLE
Add dtype and manipulation functions

### PR DESCRIPTION
**Pull request by npolina4**
RP will separate:
https://github.com/IntelPython/dpnp/pull/1582 Leverage dpctl.tensor.iinfo() and dpctl.tensor.finfo() implementation.
https://github.com/IntelPython/dpnp/pull/1583 Clean up old implementation of logic functions
https://github.com/IntelPython/dpnp/pull/1585 Updated atleast_2d and atleast_3d functions
https://github.com/IntelPython/dpnp/pull/1586 Implemented dpnp.tile function.

Add dtype and manipulation functions and tests.
Manipulation functions: dpnp.broadcast_arrays,  dpnp.repeat, dpnp.vstack, dpnp.ravel
Data Type functions: dpnp.can_cast

- [X] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?

### CODE REVIEW

File: dpnp/dpnp_algo/dpnp_algo.pyx [hunk 1]
```diff
@@ -65,7 +65,6 @@ include "dpnp_algo_counting.pxi"
 include "dpnp_algo_indexing.pxi"
 include "dpnp_algo_linearalgebra.pxi"
 include "dpnp_algo_logic.pxi"
-include "dpnp_algo_manipulation.pxi"
```
Comment (vtavana) [inline dpnp/dpnp_algo/dpnp_algo.pyx:68]: I had to remove `${CMAKE_CURRENT_SOURCE_DIR}/dpnp_algo_manipulation.pxi` from `/dpnp/dpnp_algo/CMakeLists.txt` to be able to compile the code, did not you?

I also had to remove `cpdef dpnp_descriptor dpnp_repeat(dpnp_descriptor array1, repeats, axes=*)` from `dpnp/dpnp_algo/dpnp_algo.pxd`.

File: dpnp/dpnp_iface_manipulation.py [hunk 2]
```diff
@@ -1414,20 +1506,37 @@ def unique(ar, **kwargs):
...
 
-def vstack(tup):
+def vstack(tup, *, dtype=None, casting="same_kind"):
     """
     Stack arrays in sequence vertically (row wise).
 
     For full documentation refer to :obj:`numpy.vstack`.
 
-    """
+    Examples
```
Comment (vtavana) [inline dpnp/dpnp_iface_manipulation.py:1515]: Could you please add "Limitations", "Returns", and "See Also" sections similar to `hstack`?

File: dpnp/dpnp_iface_manipulation.py [hunk 3]
```diff
@@ -1414,20 +1506,37 @@ def unique(ar, **kwargs):
     return call_origin(numpy.unique, ar, **kwargs)
 
 
-def vstack(tup):
+def vstack(tup, *, dtype=None, casting="same_kind"):
```
Comment (vtavana) [inline dpnp/dpnp_iface_manipulation.py:1509]: Could you remove `tests/test_arraymanipulation.py::TestVstack::test_generator` from skipped tests table and modify the test accordingly to be passed?

Could you also move `tests/test_random.py::TestPermutationsTestShuffle::test_shuffle1[lambda x: dpnp.vstack([x, x]).T]` to skipped test table for gpu_no_f64?

File: tests/test_arraymanipulation.py [hunk 4]
```diff
@@ -725,3 +720,336 @@ def test_results(self):
...
+
+    assert dpnp.can_cast(X, "float32") == numpy.can_cast(X_np, "float32")
+    assert dpnp.can_cast(X, dpnp.int32) == numpy.can_cast(X_np, numpy.int32)
+    assert dpnp.can_cast(X, dpnp.int64) == numpy.can_cast(X_np, numpy.int64)
+
+
+def assert_broadcast_correct(input_shapes):
+    np_arrays = [numpy.zeros(s, dtype="i1") for s in input_shapes]
+    out_np_arrays = numpy.broadcast_arrays(*np_arrays)
+    usm_arrays = [dpnp.asarray(Xnp) for Xnp in np_arrays]
```
Comment (vtavana) [inline tests/test_arraymanipulation.py:922]:
```suggestion
    dpnp_arrays = [dpnp.asarray(Xnp) for Xnp in np_arrays]
```
And rename other occurrence as well.

File: dpnp/dpnp_iface_manipulation.py [hunk 5]
```diff
@@ -293,6 +349,34 @@ def broadcast_to(array, /, shape, subok=False):
     return call_origin(numpy.broadcast_to, array, shape=shape, subok=subok)
 
 
+def can_cast(from_, to, casting="safe"):
+    """
+    Returns True if cast between data types can occur according to the casting rule.
+
+    Parameters:
```
Comment (vtavana) [inline dpnp/dpnp_iface_manipulation.py:356]:
```suggestion
    Parameters
    ----------
```
Could you do this for other similar cases as well for consistency?

File: dpnp/dpnp_iface_manipulation.py [hunk 6]
```diff
@@ -293,6 +349,34 @@ def broadcast_to(array, /, shape, subok=False):
...
+        from (dpnp_array, dtype): source data type
+        to (dtype): target data type
+        casting ({'no', 'equiv', 'safe', 'same_kind', 'unsafe'}, optional):
+            controls what kind of data casting may occur.
+
+    Returns:
+        bool: True if cast can occur according to the casting rule.
+
+    Notes
+    -----------
```
Comment (vtavana) [inline dpnp/dpnp_iface_manipulation.py:366]:
```suggestion
    Notes
    -----
```

File: dpnp/dpnp_iface_manipulation.py [hunk 7]
```diff
@@ -293,6 +349,34 @@ def broadcast_to(array, /, shape, subok=False):
...
+    If from is a scalar or array scalar, also returns True if the scalar value can
```
Comment (vtavana) [inline dpnp/dpnp_iface_manipulation.py:367]:
```suggestion
    If `from` is a scalar or array scalar, also returns `True` if the scalar value can
```

File: dpnp/dpnp_iface_manipulation.py [hunk 8]
```diff
@@ -293,6 +349,34 @@ def broadcast_to(array, /, shape, subok=False):
...
+
+    dtype_from = (
+        from_.dtype if isinstance(from_, dpnp_array) else dpnp.dtype(from_)
+    )
+    return dpt.can_cast(dtype_from, to, casting)
+
+
 def concatenate(
     arrays, /, *, axis=0, out=None, dtype=None, casting="same_kind"
 ):
```
Comment (vtavana) [inline dpnp/dpnp_iface_manipulation.py:382]: Could you modify line 398? ` dtype -> `dtype`

File: dpnp/dpnp_iface_manipulation.py [hunk 9]
```diff
@@ -847,39 +931,47 @@ def repeat(a, repeats, axis=None):
...
+
     Limitations
     -----------
-    Input array is supported as :obj:`dpnp.ndarray`.
-    Parameter ``axis`` is supported with value either ``None`` or ``0``.
-    Dimension of input array are supported to be less than ``2``.
+    Parameter `a` is supported as either :class:`dpnp.ndarray`
+    or :class:`dpctl.tensor.usm_ndarray`.
+    Otherwise ``TypeError`` exception will be raised.
     Otherwise the function will be executed sequentially on CPU.
```
Comment (vtavana) [inline dpnp/dpnp_iface_manipulation.py:957]: remove it?

**Comment by vtavana** [comment]
It seems we do not to skip the fallowing tests anymore.

tests/third_party/cupy/manipulation_tests/test_dims.py::TestBroadcast_param_0_{shapes=[(), ()]}::test_broadcast_arrays
tests/third_party/cupy/manipulation_tests/test_dims.py::TestBroadcast_param_1_{shapes=[(0,), (0,)]}::test_broadcast_arrays
tests/third_party/cupy/manipulation_tests/test_dims.py::TestBroadcast_param_2_{shapes=[(1,), (1,)]}::test_broadcast_arrays
tests/third_party/cupy/manipulation_tests/test_dims.py::TestBroadcast_param_3_{shapes=[(2,), (2,)]}::test_broadcast_arrays
tests/third_party/cupy/manipulation_tests/test_dims.py::TestBroadcast_param_4_{shapes=[(0,), (1,)]}::test_broadcast_arrays
tests/third_party/cupy/manipulation_tests/test_dims.py::TestBroadcast_param_5_{shapes=[(2, 3), (1, 3)]}::test_broadcast_arrays
tests/third_party/cupy/manipulation_tests/test_dims.py::TestBroadcast_param_6_{shapes=[(2, 1, 3, 4), (3, 1, 4)]}::test_broadcast_arrays
tests/third_party/cupy/manipulation_tests/test_dims.py::TestBroadcast_param_7_{shapes=[(4, 3, 2, 3), (2, 3)]}::test_broadcast_arrays
tests/third_party/cupy/manipulation_tests/test_dims.py::TestBroadcast_param_8_{shapes=[(2, 0, 1, 1, 3), (2, 1, 0, 0, 3)]}::test_broadcast_arrays
tests/third_party/cupy/manipulation_tests/test_dims.py::TestBroadcast_param_9_{shapes=[(0, 1, 1, 3), (2, 1, 0, 0, 3)]}::test_broadcast_arrays
tests/third_party/cupy/manipulation_tests/test_dims.py::TestBroadcast_param_10_{shapes=[(0, 1, 1, 0, 3), (5, 2, 0, 1, 0, 0, 3), (2, 1, 0, 0, 0, 3)]}::test_broadcast_arrays
tests/third_party/cupy/manipulation_tests/test_dims.py::TestInvalidBroadcast_param_0_{shapes=[(3,), (2,)]}::test_invalid_broadcast_arrays
tests/third_party/cupy/manipulation_tests/test_dims.py::TestInvalidBroadcast_param_1_{shapes=[(3, 2), (2, 3)]}::test_invalid_broadcast_arrays
tests/third_party/cupy/manipulation_tests/test_dims.py::TestInvalidBroadcast_param_2_{shapes=[(3, 2), (3, 4)]}::test_invalid_broadcast_arrays
tests/third_party/cupy/manipulation_tests/test_dims.py::TestInvalidBroadcast_param_3_{shapes=[(0,),(2,)]}::test_invalid_broadcast_arrays

File: dpnp/dpnp_iface_types.py [hunk 10]
```diff
@@ -140,6 +146,78 @@
 PZERO = numpy.PZERO
 
 
+def finfo(dtype):
```
Comment (vlad-perevezentsev) [inline dpnp/dpnp_iface_types.py:149]: Could you add tests for these new functions(`finfo()`, `isdtype()`, `iinfo()`) to increase test coverage ?

File: dpnp/dpnp_algo/dpnp_algo.pxd [hunk 11]
```diff
@@ -45,7 +45,6 @@ cdef extern from "dpnp_iface_fptr.hpp" namespace "DPNPFuncName":  # need this na
         DPNP_FN_ARGSORT
         DPNP_FN_ARGSORT_EXT
         DPNP_FN_ASTYPE
-        DPNP_FN_ASTYPE_EXT
```
Comment (vtavana) [inline dpnp/dpnp_algo/dpnp_algo.pxd:48]: Do we need to keep mapping in `dpnp/backend/kernels/dpnp_krnl_elemwise.cpp` for `DPNP_FN_ASTYPE_EXT` and `DPNP_FN_FLATTEN_EXT` or we can get rid of them?

File: dpnp/dpnp_array.py [hunk 12]
```diff
@@ -595,32 +595,37 @@ def asnumpy(self):
...
+        Returns
+        -------
+        out : dpnp.array
             If ``copy`` is False and no cast is required, then the array itself is returned.
             Otherwise, it returns a (possibly casted) copy of the array.
 
-        .. note::
-           This method currently does not support `order``, `casting``, ``copy``, and ``subok`` arguments.
-
-        .. seealso:: :meth:`numpy.ndarray.astype`
```
Comment (vtavana) [inline dpnp/dpnp_array.py:615]: Could you add examples for `astype` and `flatten`?

File: dpnp/dpnp_iface_manipulation.py [hunk 13]
```diff
@@ -1332,6 +1430,78 @@ def swapaxes(a, axis1, axis2):
     )
 
 
+def tile(a, reps):
```
Comment (vtavana) [inline dpnp/dpnp_iface_manipulation.py:1433]:
```suggestion
def tile(A, reps):
```

File: dpnp/dpnp_iface_manipulation.py [hunk 14]
```diff
@@ -1332,6 +1430,78 @@ def swapaxes(a, axis1, axis2):
...
+    """
+    Construct an array by repeating A the number of times given by reps.
```
Comment (vtavana) [inline dpnp/dpnp_iface_manipulation.py:1435]:
```suggestion
    Construct an array by repeating `A` the number of times given by reps.
```

File: dpnp/dpnp_iface_manipulation.py [hunk 15]
```diff
@@ -1332,6 +1430,78 @@ def swapaxes(a, axis1, axis2):
...
+
+    Note : Although tile may be used for broadcasting, it is strongly
+    recommended to use numpy's broadcasting operations and functions.
+
+    Parameters
+    ----------
+    A : dpnp_array
+        The input array.
+    reps : array_like
+        The number of repetitions of `a` along each axis.
```
Comment (vtavana) [inline dpnp/dpnp_iface_manipulation.py:1458]:
```suggestion
        The number of repetitions of `A` along each axis.
```

File: dpnp/dpnp_algo/dpnp_algo.pxd [hunk 16]
```diff
@@ -87,7 +86,6 @@ cdef extern from "dpnp_iface_fptr.hpp" namespace "DPNPFuncName":  # need this na
         DPNP_FN_ERF
         DPNP_FN_ERF_EXT
         DPNP_FN_EYE
-        DPNP_FN_EYE_EXT
```
Comment (vtavana) [inline dpnp/dpnp_algo/dpnp_algo.pxd:90]: I think you can also remove `DPNP_FN_EYE`, `DPNP_FN_FLATTEN`, `DPNP_FN_REPEAT`, and `DPNP_FN_REPEAT` in this file.

**Comment by ndgrigorian** [comment]
@npolina4 
With https://github.com/IntelPython/dpctl/pull/1427 you will no longer need to flatten the repeated array for the `axis=None` case. Array API spec seems to be supporting this behavior.

**Review by vtavana — APPROVED** [review]

**Comment by antonwolfy** [comment]
The below example works differently with numpy and description:
```python
a = numpy.ones((2, 3), dtype='i4')
b = a.astype('i4', copy=False)

b is a
# Out: True

a = dpnp.ones((2, 3), dtype='i4')
b = a.astype('i4', copy=False)

b is a
# Out: False
```

**Review by antonwolfy — CHANGES_REQUESTED** [review]

**Pull request closed without merging**